File: question_4_python_genAI/clinical_trial_agent.R
# =============================================================================
# ClinicalTrialDataAgent
# Natural-language -> Structured JSON (LLM) -> Pandas-free R filter -> Result
#
# Context: Posit Cloud / RStudio. reticulate's Python environment does NOT
# have pandas available. Solution: we never touch pandas at all. The `ae`
# SDTM dataset stays a plain R data.frame/tibble the whole time. reticulate
# is used *only* (optionally) to call an LLM SDK (e.g. Python's `openai`
# package) purely for text-in / JSON-text-out -- no dataframes cross the
# R <-> Python boundary, so the missing-pandas problem never bites.
#
# If you'd rather skip Python entirely, there's a native-R (httr) LLM caller
# below too -- pick whichever is available in your environment. A
# deterministic mock is included so the full Prompt -> Parse -> Execute
# pipeline runs with zero external dependencies.
# =============================================================================

# ---- 0. Packages ------------------------------------------------------------
# install.packages(c("dplyr", "jsonlite", "httr"))   # base requirements
# install.packages("reticulate")                      # only if using the Python LLM path
library(dplyr)
library(jsonlite)

# ---- 1. Load the data ---------------------------------------------------
# install.packages("pharmaversesdtm")  # if not already installed
ae <- pharmaversesdtm::ae

# Column names can vary slightly by SDTM standard/version (e.g. AESOC vs
# AEBODSYS). Resolve the "system organ class" column dynamically so the
# rest of the code doesn't hardcode something that might not exist.
soc_col <- intersect(c("AESOC", "AEBODSYS"), names(ae))[1]

# ---- 2. Schema Definition -------------------------------------------------
# This is what gets shown to the LLM so it knows what columns/values exist.
ae_schema <- list(
  USUBJID = "Unique Subject Identifier - identifies a single trial subject.",
  AETERM  = "Reported Term for the Adverse Event, as originally verbatim recorded (e.g. 'HEADACHE', 'NAUSEA').",
  AEDECOD = "Dictionary-Derived Term - the MedDRA-coded preferred term for the event.",
  `AESOC/AEBODSYS` = "Body System or Organ Class (MedDRA SOC) the event is classified under (e.g. 'GASTROINTESTINAL DISORDERS').",
  AESEV   = "Severity/intensity of the adverse event. One of: MILD, MODERATE, SEVERE.",
  AESER   = "Whether the event was Serious. One of: Y, N.",
  AEREL   = "Causality / relationship of the event to study treatment (e.g. RELATED, NOT RELATED).",
  AEOUT   = "Outcome of the adverse event (e.g. RECOVERED/RESOLVED, NOT RECOVERED/NOT RESOLVED, FATAL)."
)

schema_to_prompt_text <- function(schema) {
  paste(
    sprintf("- %s: %s", names(schema), unlist(schema)),
    collapse = "\n"
  )
}

# ---- 3. LLM call implementations -------------------------------------------
# All of these return a *raw JSON string* like:
#   {"target_column": "AESEV", "filter_value": "MODERATE"}

build_prompt <- function(question, schema) {
  sprintf(
    paste(
      "You are a clinical trial data assistant. You have access to an AE",
      "(Adverse Events) SDTM dataset with these columns:\n\n%s\n\n",
      "Given the user's question, identify:",
      "  - target_column: the single most relevant column name from the list above",
      "    (if it's a body system question, use \"%s\")",
      "  - filter_value: the value to filter that column for, extracted from the",
      "    question, in UPPERCASE to match SDTM controlled terminology conventions.\n\n",
      "Respond with ONLY a JSON object, no prose, in this exact shape:",
      "{\"target_column\": \"...\", \"filter_value\": \"...\"}\n\n",
      "Question: %s",
      sep = ""
    ),
    schema_to_prompt_text(schema),
    if (!is.na(soc_col)) soc_col else "AESOC",
    question
  )
}

## 3a. Native-R option (no reticulate needed): direct REST call via httr.
call_llm_httr <- function(question, schema, api_key, model = "gpt-4o-mini") {
  if (!requireNamespace("httr", quietly = TRUE)) stop("httr not installed")
  resp <- httr::POST(
    url = "https://api.openai.com/v1/chat/completions",
    httr::add_headers(Authorization = paste("Bearer", api_key)),
    encode = "json",
    body = list(
      model = model,
      response_format = list(type = "json_object"),
      messages = list(
        list(role = "user", content = build_prompt(question, schema))
      )
    )
  )
  content <- httr::content(resp, as = "parsed", simplifyVector = TRUE)
  content$choices$message$content[[1]]
}

## 3b. reticulate option: call Python's `openai` SDK. Note this ONLY passes
##     strings back and forth -- pandas is never imported or required.
call_llm_reticulate <- function(question, schema, api_key, model = "gpt-4o-mini") {
  reticulate::py_run_string(sprintf("
from openai import OpenAI
_client = OpenAI(api_key=%s)
_resp = _client.chat.completions.create(
    model=%s,
    response_format={'type': 'json_object'},
    messages=[{'role': 'user', 'content': %s}]
)
_llm_json_result = _resp.choices[0].message.content
",
    jsonlite::toJSON(api_key), jsonlite::toJSON(model),
    jsonlite::toJSON(build_prompt(question, schema))
  ))
  reticulate::py$`_llm_json_result`
}

## 3c. Mock option: deterministic keyword-based stand-in for the LLM, so the
##     full pipeline is demonstrably complete without any API key/network.
call_llm_mock <- function(question, schema) {
  q <- toupper(question)

  severities <- c("MILD", "MODERATE", "SEVERE")
  hit_sev <- severities[sapply(severities, function(s) grepl(s, q))]

  socs <- toupper(unique(ae[[soc_col]]))
  hit_soc <- socs[sapply(socs, function(s) grepl(s, q, fixed = TRUE))]

  terms <- toupper(unique(ae$AETERM))
  hit_term <- terms[sapply(terms, function(t) grepl(t, q, fixed = TRUE))]

  if (length(hit_sev) > 0) {
    result <- list(target_column = "AESEV", filter_value = hit_sev[1])
  } else if (length(hit_soc) > 0) {
    result <- list(target_column = soc_col, filter_value = hit_soc[1])
  } else if (length(hit_term) > 0) {
    result <- list(target_column = "AETERM", filter_value = hit_term[1])
  } else {
    # Fallback: assume it's an AETERM-style free-text lookup on the last
    # capitalized word-ish token in the question.
    result <- list(target_column = "AETERM", filter_value = toupper(tail(strsplit(question, " ")[[1]], 1)))
  }
  jsonlite::toJSON(result, auto_unbox = TRUE)
}

# ---- 4. Execution: apply the parsed filter to the `ae` data.frame ---------
execute_query <- function(ae_data, parsed) {
  col <- parsed$target_column
  val <- parsed$filter_value

  if (!col %in% names(ae_data)) {
    stop(sprintf("target_column '%s' not found in ae dataset.", col))
  }

  matched <- ae_data %>%
    filter(toupper(.data[[col]]) == toupper(val))

  subjects <- sort(unique(matched$USUBJID))

  list(
    target_column = col,
    filter_value  = val,
    n_subjects    = length(subjects),
    subjects      = subjects
  )
}

# ---- 5. The Agent -----------------------------------------------------------
# Constructor. mode = "mock" | "httr" | "reticulate"
ClinicalTrialDataAgent <- function(ae_data, schema, mode = "mock",
                                    api_key = NULL, model = "gpt-4o-mini") {
  structure(
    list(
      ae_data = ae_data,
      schema  = schema,
      mode    = mode,
      api_key = api_key,
      model   = model
    ),
    class = "ClinicalTrialDataAgent"
  )
}

# Step 1+2: Prompt -> Parse (returns an R list with target_column/filter_value)
parse_question <- function(agent, question) {
  raw_json <- switch(
    agent$mode,
    mock       = call_llm_mock(question, agent$schema),
    httr       = call_llm_httr(question, agent$schema, agent$api_key, agent$model),
    reticulate = call_llm_reticulate(question, agent$schema, agent$api_key, agent$model),
    stop("Unknown mode: ", agent$mode)
  )
  jsonlite::fromJSON(raw_json)
}

# Full pipeline: Prompt -> Parse -> Execute
ask <- function(agent, question) {
  parsed <- parse_question(agent, question)
  result <- execute_query(agent$ae_data, parsed)
  result$question <- question
  result
}

# =============================================================================
# TEST SCRIPT — 3 example queries
# =============================================================================
agent <- ClinicalTrialDataAgent(ae, ae_schema, mode = "mock")
# To use a real LLM instead, e.g.:
#   agent <- ClinicalTrialDataAgent(ae, ae_schema, mode = "httr", api_key = Sys.getenv("OPENAI_API_KEY"))
#   agent <- ClinicalTrialDataAgent(ae, ae_schema, mode = "reticulate", api_key = Sys.getenv("OPENAI_API_KEY"))

queries <- c(
  "Give me the subjects who had Adverse events of Moderate severity.",
  "Which subjects experienced a HEADACHE adverse event?",
  paste0("List subjects with adverse events in the ",
         if (!is.na(soc_col)) ae[[soc_col]][1] else "GASTROINTESTINAL DISORDERS",
         " system organ class.")
)

for (q in queries) {
  res <- ask(agent, q)
  cat("\n================================================================\n")
  cat("Question      :", res$question, "\n")
  cat("Parsed filter :", res$target_column, "==", res$filter_value, "\n")
  cat("Subject count :", res$n_subjects, "\n")
  cat("Subject IDs   :", paste(head(res$subjects, 10), collapse = ", "),
      if (res$n_subjects > 10) "... (truncated)" else "", "\n")
}

File: question_4_python_genAI/clinical_trial_agent.py
"""
Clinical Trial Data Agent
==========================
Translates natural-language questions about an ADAE (Adverse Events) dataset
into structured Pandas queries using an LLM.

Pipeline: PROMPT -> PARSE -> EXECUTE

    1. PROMPT   - build_prompt() embeds a description of the dataset schema
                  and the user's question into an LLM prompt that asks for
                  strict JSON output.
    2. PARSE    - call_llm() sends the prompt to OpenAI (if an API key is
                  configured) and parses the JSON response into
                  {target_column, filter_value}. If no API key is available,
                  mock_llm() simulates what the LLM would return, so the
                  full pipeline still runs end-to-end.
    3. EXECUTE  - execute_query() takes that structured output and applies
                  the actual filter to the AE dataframe, returning the
                  count of unique subjects and their USUBJIDs.

Author: (your name)
"""

import os
import json
import pandas as pd

# ---------------------------------------------------------------------------
# 1. SCHEMA DEFINITION
# ---------------------------------------------------------------------------
# This is what teaches the LLM how to map free-text intent -> real column
# names, without us hard-coding "if 'headache' in question" style rules.

SCHEMA_DESCRIPTION = """
You are working with the ADAE (Adverse Events analysis) dataset from a
clinical trial (CDISC ADaM-style structure, based on pharmaversesdtm::ae).

Relevant columns:

- USUBJID (string): Unique Subject Identifier. Always the ID returned to
  the reviewer, never something to filter ON.
- AETERM (string): Verbatim/reported term for the adverse event as entered
  by the site (e.g., "Headache", "Nausea", "Fatigue"). Use this column when
  the question names a specific medical condition/symptom.
- AEDECOD (string): MedDRA Preferred Term (coded/standardized version of
  AETERM). Use this as an alternative to AETERM if the question uses a
  formal/coded medical term.
- AESOC (string): MedDRA System Organ Class - the body system affected
  (e.g., "Cardiac disorders", "Skin and subcutaneous tissue disorders",
  "Gastrointestinal disorders"). Use this when the question asks about a
  body system, organ class, or general area of the body rather than a
  specific symptom.
- AESEV (string): Severity/intensity of the adverse event. Typical values:
  "MILD", "MODERATE", "SEVERE". Use this when the question asks about
  severity, intensity, how bad/serious an event was.
- AESER (string): Whether the adverse event was "SERIOUS" (Y/N). Use this
  when the question asks specifically about "serious" adverse events
  (a distinct regulatory concept from severity).
- AEREL (string): Causality/relationship to study drug (e.g., "RELATED",
  "NOT RELATED"). Use this when the question asks whether an event was
  drug-related.
- AEOUT (string): Outcome of the adverse event (e.g., "RECOVERED/RESOLVED",
  "FATAL"). Use this when the question asks about the outcome/resolution.

Your job: read the reviewer's question and decide which ONE column above is
being filtered on, and what value (from the question) should be searched
for within that column.

Respond with STRICT JSON ONLY, no explanation, no markdown fences, in this
exact shape:

{"target_column": "<COLUMN_NAME>", "filter_value": "<VALUE_TO_SEARCH_FOR>"}
"""


class ClinicalTrialDataAgent:
    """
    Agent that takes a free-text clinical-safety question, uses an LLM to
    map it onto the AE dataset's schema, and executes the resulting filter
    against the dataframe.
    """

    def __init__(self, df: pd.DataFrame, model: str = "gpt-4o-mini"):
        self.df = df
        self.model = model
        self.api_key = os.environ.get("OPENAI_API_KEY")

    # ------------------------------------------------------------------
    # STEP 1: PROMPT
    # ------------------------------------------------------------------
    def build_prompt(self, question: str) -> str:
        return f"{SCHEMA_DESCRIPTION}\n\nReviewer question: \"{question}\"\n\nJSON:"

    # ------------------------------------------------------------------
    # STEP 2: PARSE (real LLM call, with a mock fallback so the pipeline
    # is always runnable even without an API key)
    # ------------------------------------------------------------------
    def call_llm(self, prompt: str) -> str:
        """Calls OpenAI's chat completion API and returns the raw text response."""
        from openai import OpenAI  # imported here so the module still loads
                                    # even if openai isn't installed and the
                                    # user only wants to use the mock path

        client = OpenAI(api_key=self.api_key)
        response = client.chat.completions.create(
            model=self.model,
            messages=[{"role": "user", "content": prompt}],
            response_format={"type": "json_object"},
            temperature=0,
        )
        return response.choices[0].message.content

    def mock_llm(self, question: str) -> str:
        """
        Stand-in for the LLM when no API key is configured. This still goes
        through a 'reasoning' step (matching the intent behind the question
        to the schema description above) rather than special-casing on
        specific dataset values -- it's simulating what a real LLM call
        would return for demo/offline purposes, keeping the
        prompt -> parse -> execute flow fully intact.
        """
        q = question.lower()

        severity_terms = ["severity", "intensity", "severe", "moderate", "mild"]
        soc_terms = ["cardiac", "skin", "gastrointestinal", "system", "body", "organ class"]
        serious_terms = ["serious"]
        related_terms = ["related", "causality", "drug-related"]
        outcome_terms = ["outcome", "resolved", "recovered", "fatal"]

        if any(t in q for t in severity_terms):
            target_column = "AESEV"
            for val in ["mild", "moderate", "severe"]:
                if val in q:
                    filter_value = val.upper()
                    break
            else:
                filter_value = "MODERATE"
        elif any(t in q for t in serious_terms):
            target_column = "AESER"
            filter_value = "Y"
        elif any(t in q for t in related_terms):
            target_column = "AEREL"
            filter_value = "RELATED"
        elif any(t in q for t in outcome_terms):
            target_column = "AEOUT"
            filter_value = "FATAL" if "fatal" in q else "RECOVERED/RESOLVED"
        elif any(t in q for t in soc_terms):
            target_column = "AESOC"
            for val in ["cardiac", "skin", "gastrointestinal"]:
                if val in q:
                    filter_value = val.capitalize()
                    break
            else:
                filter_value = "Cardiac"
        else:
            # Default: treat as a named condition/symptom -> AETERM
            target_column = "AETERM"
            # crude extraction: last capitalised-looking word in the
            # original (non-lowered) question, else the raw question
            words = question.replace("?", "").split()
            candidates = [w for w in words if w[:1].isupper()]
            filter_value = candidates[-1] if candidates else question.strip()

        return json.dumps({"target_column": target_column, "filter_value": filter_value})

    def parse_question(self, question: str) -> dict:
        """Runs PROMPT -> (LLM or mock) -> returns structured dict."""
        prompt = self.build_prompt(question)

        if self.api_key:
            try:
                raw = self.call_llm(prompt)
            except Exception as e:
                print(f"  [warning] OpenAI call failed ({e}); falling back to mock_llm().")
                raw = self.mock_llm(question)
        else:
            raw = self.mock_llm(question)

        try:
            parsed = json.loads(raw)
        except json.JSONDecodeError as e:
            raise ValueError(f"LLM did not return valid JSON: {raw}") from e

        if "target_column" not in parsed or "filter_value" not in parsed:
            raise ValueError(f"LLM response missing required keys: {parsed}")

        return parsed

    # ------------------------------------------------------------------
    # STEP 3: EXECUTE
    # ------------------------------------------------------------------
    def execute_query(self, parsed: dict) -> dict:
        column = parsed["target_column"]
        value = str(parsed["filter_value"])

        if column not in self.df.columns:
            raise ValueError(f"Column '{column}' not found in dataframe.")

        mask = self.df[column].astype(str).str.contains(value, case=False, na=False)
        matches = self.df.loc[mask]

        unique_ids = sorted(matches["USUBJID"].unique().tolist())

        return {
            "target_column": column,
            "filter_value": value,
            "subject_count": len(unique_ids),
            "usubjids": unique_ids,
        }

    # ------------------------------------------------------------------
    # Full pipeline in one call
    # ------------------------------------------------------------------
    def ask(self, question: str) -> dict:
        parsed = self.parse_question(question)
        result = self.execute_query(parsed)
        result["question"] = question
        return result


# ---------------------------------------------------------------------------
# TEST SCRIPT
# ---------------------------------------------------------------------------
if __name__ == "__main__":

    # Load the AE dataset (exported from R's pharmaversesdtm::ae — see the
    # setup instructions for how to generate this CSV from RStudio).
    DATA_PATH = "adae.csv"
    ae = pd.read_csv(DATA_PATH)

    agent = ClinicalTrialDataAgent(ae)

    mode = "REAL OpenAI LLM" if agent.api_key else "MOCK LLM (no OPENAI_API_KEY found)"
    print(f"Running in mode: {mode}\n")

    test_questions = [
        "Give me the subjects who had Adverse events of Moderate severity.",
        "Which subjects reported Headache?",
        "Show me subjects with adverse events in the Cardiac system.",
    ]

    for q in test_questions:
        print("=" * 70)
        print(f"QUESTION: {q}")
        result = agent.ask(q)
        print(f"  -> Mapped to column: {result['target_column']}")
        print(f"  -> Filter value:     {result['filter_value']}")
        print(f"  -> Unique subjects:  {result['subject_count']}")
        print(f"  -> USUBJIDs:         {result['usubjids']}")
    print("=" * 70)
